File: 2.3-2.py
def sort(array):
    n = len(array)

    if(n>1):
        half = int(n/2)
        left = [0]*half
        right = [0]*(n-half)

        for i in range(half):
            left[i] = array[i]
        for i in range(n-half):
            right[i] = array[half+i]

        left = sort(left)
        right = sort(right)
        j = 0
        k = 0

        for i in range(n):

            if(k == (n-half)):
                array[i] = left[j]
                j += 1
            elif(j == half):
                array[i] = right[k]
                k += 1
            elif(left[j]<right[k]):
                array[i] = left[j]
                j += 1
            else:
                array[i] = right[k]
                k += 1

    return array
    
array = [64,73,85,72,7,3,6,3,2,5,3,4,5,5,3,2,5,0,7,5,0,2,6,7,0,7,6,6755,46,24,25,26,78,93]
array = sort(array)
print(len(array), array)

File: Chapter-2/Exercises/2.3-2.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(int arr[], int p, int q, int r) {
    int n1{q - p + 1}, n2{r - q};
    int left[n1] = {}, right[n2] = {};

    for(int i{}; i<n1; i++)
        left[i] = arr[p + i];
    for(int i{}; i<n2; i++)
        right[i] = arr[q + 1 + i];

    int i{0}, j{0}, k{p};
    while(i<n1 && j<n2) {
        if(left[i]<right[j]) {
            arr[k] = left[i];
            i++, k++;
        }
        else {
            arr[k] = right[j];
            j++, k++;
        };
    };

    while(i<n1) {
        arr[k] = left[i];
        i++, k++;
    };
    while(j<n2) {
        arr[k] = right[j];
        j++, k++;
    };
}

void mergeSort(int arr[], int p, int r) {
    if(p<r){
        int q = (p+r)/2;
        mergeSort(arr, p, q);
        mergeSort(arr, q+1, r);
        merge(arr, p, q, r);
    };
}

int main() {
    int arr[11] = {9,8,7,6,5,4,3,2,2,1,0};
    int l = sizeof(arr)/sizeof(arr[0]);
    mergeSort(arr,0,l-1);

    for(int x: arr)
        cout<<x<<" ";
    return 0;
}
